### PR TITLE
Fixing Re-renders

### PR DESCRIPTION
**Pull request by karanvirsb**
Added memoization

### CODE REVIEW

**Comment by alexdanilowicz** [comment]
Just curious @karanvirsb, what's the reasoning for the `useStore` needing to be split out for it to memo-ize?

**Comment by karanvirsb** [comment]
> Just curious @karanvirsb, what's the reasoning for the `useStore` needing to be split out for it to memo-ize?
 
TLDR - Better practice and less renders.
 
It is not needed for memoizing but rather for fixing re-renders. Due to it all depending on global state rather than individual ones it would cause renders. 

Here is a video I watched from Jack Herrington: https://www.youtube.com/watch?v=aOt4Hz3ze3Q&ab_channel=JackHerrington

He explains and demonstrates it better .

**Comment by alexdanilowicz** [comment]
> > Just curious @karanvirsb, what's the reasoning for the `useStore` needing to be split out for it to memo-ize?
> 
> TLDR - Better practice and less renders.
> 
> It is not needed for memoizing but rather for fixing re-renders. Due to it all depending on global state rather than individual ones it would cause renders.
> 
> Here is a video I watched from Jack Herrington: https://www.youtube.com/watch?v=aOt4Hz3ze3Q&ab_channel=JackHerrington
> 
> He explains and demonstrates it better .

Oh nice, makes sense! Totally aligned.

Whenever this is ready feel free to mark it not as a draft, and will merge!

**Comment by karanvirsb** [comment]
No problem will do @alexdanilowicz

File: packages/core/src/components/ColorPalette/ColorDisplay.tsx [hunk 1]
```diff
@@ -54,8 +54,8 @@ function MirrorfulMenuButton({
     />
   )
 }
-
-export function ColorDisplay({
+const MotionRow = motion(VariantRow)
```
Comment (alexdanilowicz) [inline packages/core/src/components/ColorPalette/ColorDisplay.tsx:57]: i don't think this works actually. I no longer see the animation. Could we just add back in `<motion.div>` top level here?

Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/ColorDisplay.tsx]: Thank you for pointing that out Alex, I have added back `<motion.div>`.

File: packages/core/src/components/ColorPalette/ColorDisplay.tsx [hunk 2]
```diff
@@ -224,7 +224,7 @@ export function ColorDisplay({
             }
             return (
               <motion.div
-                key={variant.name}
+                key={`${colorName}-${variant.name}`}
```
[the remark this answers is not in the document]
Comment (alexdanilowicz) [reply packages/core/src/components/ColorPalette/ColorDisplay.tsx:227]: because the hsl affects all other colors

keys might be a red herring tho

File: packages/core/src/components/ColorPalette/VariantRow.tsx [hunk 3]
```diff
@@ -111,6 +112,31 @@ export function VariantRow({
...
+            },
+          },
+          isBase
+        )
+      }, 250)
+    }
+    return () => {
+      clearTimeout(timer)
+    }
+  }, [color])
```
[the remark this answers is not in the document]
Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/VariantRow.tsx:138]: @Teddarific  and @alexdanilowicz I did this to debounce the updates. So that while the user is changing the slider it won't update. I can try to remove it and see if I can do something else.

File: packages/core/src/components/ColorPalette/ColorDisplay.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/ColorDisplay.tsx:227]: The key does not have the hex code in it anymore rather the name of the color name and variant name. Meaning like Purple-50, Purple-100 ... . It will only should the animation once due to display colors being memoized.

File: packages/core/src/components/ColorPalette/VariantRow.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexdanilowicz) [reply packages/core/src/components/ColorPalette/VariantRow.tsx:138]: I don't think debouncing is the right approach. I don't like relying on arbitrary time to fix something. I think we should remove the timers. I'm cool with merging this PR without the timers and we fix the HSL issue separately. 


Want to remove the timers, we get this in a mergeable state, and we go from there? @karanvirsb 

cc @Teddarific

File: packages/core/src/components/ColorPalette/VariantRow.tsx [hunk 4]
```diff
@@ -152,7 +176,7 @@ export function VariantRow({
         css={{
           height: '3rem',
           width: '100%',
-          backgroundColor: `${token.value}`,
+          backgroundColor: `${changing ? color : variant.token.value}`,
```
Comment (alexdanilowicz) [inline packages/core/src/components/ColorPalette/VariantRow.tsx:179]: Let's also remove the concept of the changing var @karanvirsb

File: packages/core/src/components/ColorPalette/VariantRow.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/VariantRow.tsx:138]: Hey @alexdanilowicz if we remove the timer then it will cause a render every time the slider changes.

[the remark this answers is not in the document]
Comment (alexdanilowicz) [reply packages/core/src/components/ColorPalette/VariantRow.tsx:138]: moving convo to: https://mirrorful.slack.com/archives/C04RR2D7C65/p1684188510158349?thread_ts=1684125237.982739&cid=C04RR2D7C65

[the remark this answers is not in the document]
Comment (Pranav2612000) [reply packages/core/src/components/ColorPalette/VariantRow.tsx:138]: How about using `onChangeEnd`  [ [ref](https://chakra-ui.com/docs/components/slider#getting-the-final-value-when-dragging-the-slider) ]

File: packages/core/src/components/ColorPalette/VariantRow.tsx [hunk 5]
```diff
@@ -134,16 +160,14 @@ export function VariantRow({
         l: val / 100,
       }).toHexString()
     }
-
-    const newToken = {
-      ...variant.token,
-      value: newColor,
-    }
-    onUpdateVariant({ ...variant, token: newToken }, isBase)
+    setColor(() => newColor)
```
Comment (Pranav2612000) [inline packages/core/src/components/ColorPalette/VariantRow.tsx:163]: Nit: Why not just do 
```suggestion
    setColor(newColor)
```

?

Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/VariantRow.tsx]: Yeah removed it. Its usually for lazy initialization. Was using it to test something. Thanks!

File: packages/core/src/components/ColorPalette/ColorDisplay.tsx [hunk 6]
```diff
@@ -338,3 +341,7 @@ export function ColorDisplay({
     </Box>
   )
 }
+
+export const ColorDisplay = React.memo(ColorsDisplay, (prev, next) => {
+  return prev.colorData === next.colorData
```
[the remark this answers is not in the document]
Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/ColorDisplay.tsx]: It does not sometimes you are right! Made a change to check between the old and new props!

File: packages/core/src/components/ColorPalette/VariantRow.tsx (hunk 4 above)
Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/VariantRow.tsx:177]: I cannot remove this otherwise the change will only display after the user has stopped changing.

File: packages/core/src/components/ColorPalette/VariantRow.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (karanvirsb) [reply packages/core/src/components/ColorPalette/VariantRow.tsx]: Yeah used onChangeEnd and onChangeStart to tackle the timer. Thanks for the suggestion @Pranav2612000 !
